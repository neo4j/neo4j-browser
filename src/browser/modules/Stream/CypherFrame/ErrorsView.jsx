/*
 * Copyright (c) 2002-2019 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { Component } from 'react'
import { withBus } from 'react-suber'
import Ellipsis from 'browser-components/Ellipsis'
import {
  ExclamationTriangleIcon,
  PlayIcon
} from 'browser-components/icons/Icons'
import { deepEquals } from 'services/utils'
import Render from 'browser-components/Render'
import {
  executeCommand,
  listDbsCommand
} from 'shared/modules/commands/commandsDuck'
import { listAvailableProcedures } from 'shared/modules/cypher/procedureFactory'
import {
  isUnknownProcedureError,
  isNoDbAccessError
} from 'services/cypherErrorsHelper'
import { errorMessageFormater } from './../errorMessageFormater'

import {
  StyledCypherErrorMessage,
  StyledHelpContent,
  StyledErrorH4,
  StyledPreformattedArea,
  StyledHelpDescription,
  StyledDiv,
  StyledLink,
  StyledLinkContainer,
  StyledHelpFrame,
  ErrorText
} from '../styled'

export class ErrorsView extends Component {
  shouldComponentUpdate(props, state) {
    return !deepEquals(props.result, this.props.result)
  }

  render() {
    const { result: error, bus } = this.props
    if (!error || !error.code) {
      return null
    }
    const fullError = errorMessageFormater(null, error.message)

    return (
      <StyledHelpFrame>
        <StyledHelpContent>
          <StyledHelpDescription>
            <StyledCypherErrorMessage>ERROR</StyledCypherErrorMessage>
            <StyledErrorH4>{error.code}</StyledErrorH4>
          </StyledHelpDescription>
          <StyledDiv>
            <StyledPreformattedArea>{fullError.message}</StyledPreformattedArea>
          </StyledDiv>
          <Render if={isUnknownProcedureError(error)}>
            <StyledLinkContainer>
              <StyledLink
                onClick={() => onItemClick(bus, listAvailableProcedures)}
              >
                <PlayIcon />
                &nbsp;List available procedures
              </StyledLink>
            </StyledLinkContainer>
          </Render>
          <Render if={isNoDbAccessError(error)}>
            <StyledLinkContainer>
              <StyledLink
                onClick={() => onItemClick(bus, `:${listDbsCommand}`)}
              >
                <PlayIcon />
                &nbsp;List available databases
              </StyledLink>
            </StyledLinkContainer>
          </Render>
        </StyledHelpContent>
      </StyledHelpFrame>
    )
  }
}

const onItemClick = (bus, statement) => {
  const action = executeCommand(statement)
  bus.send(action.type, action)
}

export const ErrorsViewBus = withBus(ErrorsView)

export class ErrorsStatusbar extends Component {
  shouldComponentUpdate(props, state) {
    return !deepEquals(props.result, this.props.result)
  }

  render() {
    const error = this.props.result
    if (!error || (!error.code && !error.message)) return null
    const fullError = errorMessageFormater(error.code, error.message)

    return (
      <Ellipsis>
        <ErrorText title={fullError.title}>
          <ExclamationTriangleIcon /> {fullError.message}
        </ErrorText>
      </Ellipsis>
    )
  }
}
