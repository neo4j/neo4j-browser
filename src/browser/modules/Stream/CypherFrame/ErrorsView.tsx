/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { Component } from 'react'
import { withBus } from 'react-suber'
import Ellipsis from 'browser-components/Ellipsis'
import {
  ExclamationTriangleIcon,
  PlayIcon
} from 'browser-components/icons/Icons'
import { deepEquals } from 'services/utils'

import {
  commandSources,
  executeCommand,
  listDbsCommand
} from 'shared/modules/commands/commandsDuck'
import { listAvailableProcedures } from 'shared/modules/cypher/procedureFactory'
import {
  isUnknownProcedureError,
  isNoDbAccessError,
  isPeriodicCommitError
} from 'services/cypherErrorsHelper'
import { errorMessageFormater } from './../errorMessageFormater'

import {
  StyledCypherErrorMessage,
  StyledHelpContent,
  StyledErrorH4,
  StyledPreformattedArea,
  StyledHelpDescription,
  StyledDiv,
  StyledLink,
  StyledLinkContainer,
  StyledHelpFrame,
  ErrorText
} from '../styled'

export class ErrorsView extends Component<any> {
  shouldComponentUpdate(props: any) {
    return !deepEquals(props.result, this.props.result)
  }

  render() {
    const { result: error, bus } = this.props
    if (!error || !error.code) {
      return null
    }
    const fullError = errorMessageFormater(null, error.message)

    return (
      <StyledHelpFrame>
        <StyledHelpContent>
          <StyledHelpDescription>
            <StyledCypherErrorMessage>ERROR</StyledCypherErrorMessage>
            <StyledErrorH4>{error.code}</StyledErrorH4>
          </StyledHelpDescription>
          <StyledDiv>
            <StyledPreformattedArea>{fullError.message}</StyledPreformattedArea>
          </StyledDiv>
          {isUnknownProcedureError(error) && (
            <StyledLinkContainer>
              <StyledLink
                onClick={() => onItemClick(bus, listAvailableProcedures)}
              >
                <PlayIcon />
                &nbsp;List available procedures
              </StyledLink>
            </StyledLinkContainer>
          )}
          {isNoDbAccessError(error) && (
            <StyledLinkContainer>
              <StyledLink
                onClick={() => onItemClick(bus, `:${listDbsCommand}`)}
              >
                <PlayIcon />
                &nbsp;List available databases
              </StyledLink>
            </StyledLinkContainer>
          )}
          {isPeriodicCommitError(error) && (
            <StyledLinkContainer>
              <StyledLink onClick={() => onItemClick(bus, `:help auto`)}>
                <PlayIcon />
                &nbsp;Info on the <code>:auto</code> command
              </StyledLink>
              &nbsp;(auto-committing transactions)
            </StyledLinkContainer>
          )}
        </StyledHelpContent>
      </StyledHelpFrame>
    )
  }
}

const onItemClick = (bus: any, statement: any) => {
  const action = executeCommand(statement, { source: commandSources.button })
  bus.send(action.type, action)
}

export const ErrorsViewBus = withBus(ErrorsView)

export class ErrorsStatusbar extends Component<any> {
  shouldComponentUpdate(props: any) {
    return !deepEquals(props.result, this.props.result)
  }

  render() {
    const error = this.props.result
    if (!error || (!error.code && !error.message)) return null
    const fullError = errorMessageFormater(error.code, error.message)

    return (
      <Ellipsis>
        <ErrorText title={fullError.title}>
          <ExclamationTriangleIcon /> {fullError.message}
        </ErrorText>
      </Ellipsis>
    )
  }
}
