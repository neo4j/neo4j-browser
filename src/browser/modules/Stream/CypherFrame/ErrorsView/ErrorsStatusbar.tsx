/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import React, { Component } from 'react'

import { deepEquals, ExclamationTriangleIcon } from 'nxComps/common'

import Ellipsis from 'browser-components/Ellipsis'
import { errorMessageFormater } from 'project-root/src/browser/modules/Stream/errorMessageFormater'
import { ErrorText } from 'project-root/src/browser/modules/Stream/styled'
import { BrowserRequestResult } from 'project-root/src/shared/modules/requests/requestsDuck'
import { BrowserError } from 'services/exceptions'

type ErrorsStatusBarProps = {
  result: BrowserRequestResult
}
export class ErrorsStatusbar extends Component<ErrorsStatusBarProps> {
  shouldComponentUpdate(props: ErrorsStatusBarProps): boolean {
    return !deepEquals(props.result, this.props.result)
  }

  render(): null | JSX.Element {
    const error = this.props.result as BrowserError
    if (!error || (!error.code && !error.message)) return null
    const fullError = errorMessageFormater(error.code, error.message)

    return (
      <Ellipsis>
        <ErrorText title={fullError.title}>
          <ExclamationTriangleIcon /> {fullError.message}
        </ErrorText>
      </Ellipsis>
    )
  }
}
