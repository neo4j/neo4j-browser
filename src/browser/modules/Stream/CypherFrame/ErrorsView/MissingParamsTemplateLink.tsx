/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react'

import {
  StyledMissingParamsTemplateLink,
  StyledParamsTemplateClickableArea,
  StyledSpecifyParamsText
} from 'project-root/src/browser/modules/Stream/styled'
import { stringModifier } from 'services/bolt/cypherTypesFormatting'
import { BrowserError } from 'services/exceptions'
import { stringifyMod } from 'services/utils'

type MissingParamsTemplateLinkProps = {
  error: BrowserError
  params: Record<string, unknown>
  setEditorContent: (cmd: string) => void
  onTemplateHelpMessageClick: () => void
}
export const MissingParamsTemplateLink = ({
  setEditorContent,
  params,
  error,
  onTemplateHelpMessageClick
}: MissingParamsTemplateLinkProps): JSX.Element => {
  const handleTemplateHelpMessageClick = (): void => {
    onTemplateHelpMessageClick()
    const missingParams = getMissingParams(error.message)
    const template = getSettingMissingParamsTemplate(missingParams, params)
    setEditorContent(template)
  }

  const getMissingParams = (missingParamsErrorMessage: string) => {
    const regExp = new RegExp(`^Expected parameter\\(s\\): (.*?)$`, 'g')
    const match = regExp.exec(missingParamsErrorMessage)
    if (match && match.length > 1) {
      return match[1].split(', ')
    } else {
      return []
    }
  }
  const getSettingMissingParamsTemplate = (
    missingParams: string[],
    existingParams: Record<string, unknown>
  ): string => {
    const missingParamsTemplate = missingParams
      .map(param => `  "${param}": fill_in_your_value`)
      .join(',\n')

    let existingParamsTemplate = ''
    const existingParamsIsEmpty = Object.keys(existingParams).length === 0
    if (!existingParamsIsEmpty) {
      const existingParamsStringWithBracketsAndSurroundingNewlines =
        stringifyMod(existingParams, stringModifier, true)
      const existingParamsStringCleaned =
        existingParamsStringWithBracketsAndSurroundingNewlines
          .slice(
            1,
            existingParamsStringWithBracketsAndSurroundingNewlines.length - 1
          )
          .trim()
      existingParamsTemplate = `,\n\n  ${existingParamsStringCleaned}`
    }

    return `:params \n{\n${missingParamsTemplate}${existingParamsTemplate}\n}`
  }

  return (
    <StyledMissingParamsTemplateLink>
      <span>Use this template to add missing parameter(s):</span>
      <StyledParamsTemplateClickableArea
        onClick={() => handleTemplateHelpMessageClick()}
        aria-label={
          'Set editor content with template to be used for setting the missing parameters'
        }
      >
        :params{' {'}
        <StyledSpecifyParamsText>specify params</StyledSpecifyParamsText>
        {'}'}
      </StyledParamsTemplateClickableArea>
    </StyledMissingParamsTemplateLink>
  )
}
