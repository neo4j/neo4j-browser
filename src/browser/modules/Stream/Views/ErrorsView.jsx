/*
 * Copyright (c) 2002-2017 "Neo Technology,"
 * Network Engine for Objects in Lund AB [http://neotechnology.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { withBus } from 'preact-suber'
import { executeCommand } from 'shared/modules/commands/commandsDuck'
import { listAvailableProcedures } from 'shared/modules/cypher/procedureFactory'
import { isUnknownProcedureError } from 'services/cypherErrorsHelper'
import Visible from 'browser-components/Visible'
import { PlayIcon } from 'browser-components/icons/Icons'
import {
  StyledCypherErrorMessage,
  StyledHelpContent,
  StyledH4,
  StyledPreformattedArea,
  StyledHelpDescription,
  StyledDiv,
  StyledLink,
  StyledLinkContainer,
  StyledHelpFrame
} from '../styled'

const ErrorsView = ({error, style, bus}) => {
  if (!error) {
    return null
  }
  return (
    <StyledHelpFrame style={style}>
      <StyledHelpContent>
        <StyledHelpDescription>
          <StyledCypherErrorMessage>ERROR</StyledCypherErrorMessage>
          <StyledH4>{error.code}</StyledH4>
        </StyledHelpDescription>
        <StyledDiv>
          <StyledPreformattedArea>{error.message}</StyledPreformattedArea>
        </StyledDiv>
        <Visible if={isUnknownProcedureError(error)}>
          <StyledLinkContainer>
            <StyledLink onClick={() => onItemClick(bus)}><PlayIcon />&nbsp;List available procedures</StyledLink>
          </StyledLinkContainer>
        </Visible>
      </StyledHelpContent>
    </StyledHelpFrame>)
}

const onItemClick = (bus) => {
  const action = executeCommand(listAvailableProcedures)
  bus.send(action.type, action)
}

export default withBus(ErrorsView)
