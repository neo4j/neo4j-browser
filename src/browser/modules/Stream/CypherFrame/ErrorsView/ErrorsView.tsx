/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import React, { Component } from 'react'
import { connect } from 'react-redux'
import { withBus } from 'react-suber'
import { Action, Dispatch } from 'redux'
import { Bus } from 'suber'

import { PlayIcon } from 'browser-components/icons/LegacyIcons'

import {
  StyledCypherErrorMessage,
  StyledDiv,
  StyledErrorH4,
  StyledHelpContent,
  StyledHelpDescription,
  StyledHelpFrame,
  StyledLink,
  StyledLinkContainer,
  StyledPreformattedArea
} from '../../styled'
import { MissingParamsTemplateLink } from './MissingParamsTemplateLink'
import { GlobalState } from 'project-root/src/shared/globalState'
import {
  commandSources,
  executeCommand,
  listDbsCommand
} from 'project-root/src/shared/modules/commands/commandsDuck'
import { getListProcedureQuery } from 'shared/modules/cypher/functionsAndProceduresHelper'
import * as editor from 'project-root/src/shared/modules/editor/editorDuck'
import { getParams } from 'project-root/src/shared/modules/params/paramsDuck'
import { BrowserRequestResult } from 'project-root/src/shared/modules/requests/requestsDuck'
import { GENERATE_SET_MISSING_PARAMS_TEMPLATE } from 'project-root/src/shared/modules/udc/udcDuck'
import {
  isImplicitTransactionError,
  isNoDbAccessError,
  isParameterMissingError,
  isUnknownProcedureError
} from 'services/cypherErrorsHelper'
import { BrowserError } from 'services/exceptions'
import { deepEquals } from 'neo4j-arc/common'
import { getSemanticVersion } from 'shared/modules/dbMeta/dbMetaDuck'
import { SemVer } from 'semver'
import {
  flattenAndInvertErrors,
  formatError,
  FormattedError
} from '../errorUtils'
import { gqlErrorsAndNotificationsEnabled } from 'services/gqlUtils'
import styled from 'styled-components'

const StyledErrorsViewInnerComponentContent = styled.div<{ nested: boolean }>`
  padding-left: ${props => (props.nested ? '20px' : '0')};
`

type ErrorsViewInnerProps = {
  formattedError: FormattedError
  nested?: boolean
}

class ErrorsViewInnerComponent extends Component<ErrorsViewInnerProps> {
  render(): null | JSX.Element {
    const { formattedError, nested = false } = this.props

    return (
      <StyledErrorsViewInnerComponentContent nested={nested}>
        <StyledHelpDescription>
          {!nested && (
            <React.Fragment>
              <StyledCypherErrorMessage>ERROR</StyledCypherErrorMessage>
              <StyledErrorH4>{formattedError.title}</StyledErrorH4>
            </React.Fragment>
          )}
          {nested && <h6>{formattedError.title}</h6>}
        </StyledHelpDescription>
        {formattedError.description && (
          <StyledDiv>
            <StyledPreformattedArea data-testid={'cypherFrameErrorMessage'}>
              {formattedError?.description}
            </StyledPreformattedArea>
          </StyledDiv>
        )}
      </StyledErrorsViewInnerComponentContent>
    )
  }
}

export type ErrorsViewProps = {
  result: BrowserRequestResult
  bus: Bus
  neo4jVersion: SemVer | null
  params: Record<string, unknown>
  executeCmd: (cmd: string) => void
  setEditorContent: (cmd: string) => void
  depth?: number
  gqlErrorsAndNotificationsEnabled?: boolean
}

type ErrorsViewState = {
  nestedErrorsToggled: boolean
}

class ErrorsViewComponent extends Component<ErrorsViewProps, ErrorsViewState> {
  state = {
    nestedErrorsToggled: false
  }

  shouldComponentUpdate(
    props: ErrorsViewProps,
    state: ErrorsViewState
  ): boolean {
    return (
      !deepEquals(props.result, this.props.result) ||
      !deepEquals(props.params, this.props.params) ||
      !deepEquals(state, this.state)
    )
  }

  render(): null | JSX.Element {
    const {
      bus,
      params,
      executeCmd,
      setEditorContent,
      neo4jVersion,
      gqlErrorsAndNotificationsEnabled = false
    } = this.props

    const error = this.props.result as BrowserError

    const invertedErrors = flattenAndInvertErrors(
      error,
      gqlErrorsAndNotificationsEnabled
    )
    const [deepestError] = invertedErrors
    const nestedErrors = invertedErrors.slice(1)
    const togglable = nestedErrors.length > 0
    const setNestedErrorsToggled = (toggled: boolean) => {
      this.setState({
        nestedErrorsToggled: toggled
      })
    }

    if (!error) {
      return null
    }

    const formattedError = formatError(
      deepestError,
      gqlErrorsAndNotificationsEnabled
    )

    if (!formattedError?.title) {
      return null
    }

    const handleSetMissingParamsTemplateHelpMessageClick = () => {
      bus.send(GENERATE_SET_MISSING_PARAMS_TEMPLATE, undefined)
    }

    return (
      <StyledHelpFrame>
        <StyledHelpContent>
          <ErrorsViewInnerComponent formattedError={formattedError} />
          {isUnknownProcedureError(error) && (
            <StyledLinkContainer>
              <StyledLink
                onClick={() => executeCmd(getListProcedureQuery(neo4jVersion))}
              >
                <PlayIcon />
                &nbsp;List available procedures
              </StyledLink>
            </StyledLinkContainer>
          )}
          {isNoDbAccessError(error) && (
            <StyledLinkContainer>
              <StyledLink onClick={() => executeCmd(`:${listDbsCommand}`)}>
                <PlayIcon />
                &nbsp;List available databases
              </StyledLink>
            </StyledLinkContainer>
          )}
          {isImplicitTransactionError(error) && (
            <StyledLinkContainer>
              <StyledLink onClick={() => executeCmd(`:help auto`)}>
                <PlayIcon />
                &nbsp;Info on the <code>:auto</code> command
              </StyledLink>
              &nbsp;(auto-committing transactions)
            </StyledLinkContainer>
          )}
          {isParameterMissingError(error) && (
            <MissingParamsTemplateLink
              error={error}
              params={params}
              setEditorContent={setEditorContent}
              onTemplateHelpMessageClick={
                handleSetMissingParamsTemplateHelpMessageClick
              }
            />
          )}
          {togglable && (
            <StyledLinkContainer>
              <StyledLink
                onClick={() =>
                  setNestedErrorsToggled(!this.state.nestedErrorsToggled)
                }
              >
                &nbsp;
                {this.state.nestedErrorsToggled ? 'Show less' : 'Show more'}
              </StyledLink>
            </StyledLinkContainer>
          )}
          {this.state.nestedErrorsToggled &&
            nestedErrors.map((nestedError, index) => (
              <ErrorsViewInnerComponent
                key={index}
                nested={true}
                formattedError={formatError(
                  nestedError,
                  gqlErrorsAndNotificationsEnabled
                )}
              />
            ))}
        </StyledHelpContent>
      </StyledHelpFrame>
    )
  }
}

const mapStateToProps = (state: GlobalState) => ({
  params: getParams(state),
  neo4jVersion: getSemanticVersion(state),
  gqlErrorsAndNotificationsEnabled: gqlErrorsAndNotificationsEnabled(state)
})

const mapDispatchToProps = (
  _dispatch: Dispatch<Action>,
  ownProps: ErrorsViewProps
) => {
  return {
    executeCmd: (cmd: string) => {
      const action = executeCommand(cmd, {
        source: commandSources.button
      })
      ownProps.bus.send(action.type, action)
    },
    setEditorContent: (cmd: string) => {
      ownProps.bus.send(editor.SET_CONTENT, editor.setContent(cmd))
    }
  }
}

export const ErrorsView = withBus(
  connect(mapStateToProps, mapDispatchToProps)(ErrorsViewComponent)
)
