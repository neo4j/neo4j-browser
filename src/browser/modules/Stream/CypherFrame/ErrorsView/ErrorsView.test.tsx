/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import { render } from '@testing-library/react'
import React from 'react'
import { createBus } from 'suber'

import { ErrorsView, ErrorsViewProps } from './ErrorsView'
import { BrowserError } from 'services/exceptions'
import { Provider } from 'react-redux'
import { initialState as initialMetaState } from 'shared/modules/dbMeta/dbMetaDuck'
import { initialState as initialSettingsState } from 'shared/modules/settings/settingsDuck'

const withProvider = (store: any, children: any) => {
  return <Provider store={store}>{children}</Provider>
}

const mount = (props: Partial<ErrorsViewProps>, state?: any) => {
  const defaultProps: ErrorsViewProps = {
    result: null,
    bus: createBus(),
    params: {},
    executeCmd: jest.fn(),
    setEditorContent: jest.fn(),
    neo4jVersion: null
  }

  const combinedProps = {
    ...defaultProps,
    ...props
  }

  const initialState = {
    meta: initialMetaState,
    settings: initialSettingsState
  }

  const combinedState = { ...initialState, ...state }

  const store = {
    subscribe: () => {},
    dispatch: () => {},
    getState: () => ({
      ...combinedState
    })
  }

  return render(withProvider(store, <ErrorsView {...combinedProps} />))
}

describe('ErrorsView', () => {
  test('displays nothing if no errors', () => {
    // Given
    const props = {
      result: null
    }

    // When
    const { container } = mount(props)

    // Then
    expect(container).toMatchSnapshot()
  })

  test('does display an error', () => {
    // Given
    const error: BrowserError = {
      code: 'Test.Error',
      message: 'Test error description',
      type: 'Neo4jError'
    }
    const props = {
      result: error
    }

    // When
    const { container } = mount(props)

    // Then
    expect(container).toMatchSnapshot()
  })

  test('does display an error for GQL status codes', () => {
    // Given
    const error: BrowserError = {
      code: 'Test.Error',
      message: 'Test error description',
      type: 'Neo4jError',
      gqlStatus: '22N14',
      gqlStatusDescription:
        "error: data exception - invalid temporal value combination. Cannot select both epochSeconds and 'datetime'.",
      cause: undefined
    }

    const props = {
      result: error
    }

    const state = {
      meta: {
        server: {
          version: '5.27.0'
        }
      }
    }

    // When
    const { container } = mount(props, state)

    // Then
    expect(container).toMatchSnapshot()
  })

  test('does display a nested error for GQL status codes', () => {
    // Given
    const error: BrowserError = {
      code: 'Test.Error',
      message: 'Test error description',
      type: 'Neo4jError',
      gqlStatus: '42N51',
      gqlStatusDescription:
        'error: syntax error or access rule violation - invalid parameter. Invalid parameter $`param`. ',
      cause: {
        gqlStatus: '22G03',
        gqlStatusDescription: 'error: data exception - invalid value type',
        cause: {
          gqlStatus: '22N27',
          gqlStatusDescription:
            "error: data exception - invalid entity type. Invalid input '******' for $`param`. Expected to be STRING.",
          cause: undefined
        }
      }
    }

    const props = {
      result: error
    }

    const state = {
      meta: {
        server: {
          version: '5.27.0'
        }
      }
    }

    // When
    const { container } = mount(props, state)

    // Then
    expect(container).toMatchSnapshot()
  })

  test('displays procedure link if unknown procedure', () => {
    // Given
    const error: BrowserError = {
      code: 'Neo.ClientError.Procedure.ProcedureNotFound',
      message: 'not found',
      type: 'Neo4jError'
    }
    const props = {
      result: error
    }

    // When
    const { container, getByText } = mount(props)

    // Then
    expect(container).toMatchSnapshot()
    expect(getByText('List available procedures')).not.toBeUndefined()
  })

  test('displays procedure link if periodic commit error', () => {
    // Given
    const error: BrowserError = {
      code: 'Neo.ClientError.Statement.SemanticError',
      message:
        'Executing queries that use periodic commit in an open transaction is not possible.',
      type: 'Neo4jError'
    }
    const props = {
      result: error
    }

    // When
    const { getByText } = mount(props)

    // Then
    // We need to split up because of the use of <code> tags in the rendered document
    expect(getByText(/info on the/i)).not.toBeUndefined()
    expect(getByText(':auto')).not.toBeUndefined()
    expect(getByText('(auto-committing transactions)')).not.toBeUndefined()
  })
})
