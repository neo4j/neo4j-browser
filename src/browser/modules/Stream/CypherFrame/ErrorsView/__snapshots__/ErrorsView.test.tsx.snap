// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorsView displays nothing if no errors 1`] = `<div />`;

exports[`ErrorsView displays procedure link if unknown procedure 1`] = `
<div>
  <div
    class="sc-kfPuZi fgEQGn"
  >
    <div
      class="sc-fKVqWL hFvVsk"
    >
      <div
        class="sc-gUQvok ilTLcb"
      >
        <div
          class="sc-bBHxTw cOKOWM"
        >
          <div
            class="sc-iJKOTD sc-ezbkAF hhmYSB fuiMKG"
          >
            ERROR
          </div>
          <h4
            class="sc-hiCibw HoeQn"
          >
            Neo.ClientError.Procedure.ProcedureNotFound
          </h4>
        </div>
        <div
          class="sc-iwjdpV"
        >
          <pre
            class="sc-cCcXHH haCMIe"
            data-testid="cypherFrameErrorMessage"
          >
            not found
          </pre>
        </div>
      </div>
      <div
        class="sc-llYSUQ dMBrER"
      >
        <a
          class="sc-cxpSdN gYcTYD"
        >
          <i
            class="sc-bdvvtL jWhaJw fa fa-play-circle-o"
            style="line-height: inherit;"
          />
           List available procedures
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`ErrorsView does display a nested error for GQL status codes 1`] = `
<div>
  <div
    class="sc-kfPuZi fgEQGn"
  >
    <div
      class="sc-fKVqWL hFvVsk"
    >
      <div
        class="sc-gUQvok ilTLcb"
      >
        <div
          class="sc-bBHxTw cOKOWM"
        >
          <div
            class="sc-iJKOTD sc-ezbkAF hhmYSB fuiMKG"
          >
            ERROR
          </div>
          <h4
            class="sc-hiCibw HoeQn"
          >
            22N27: Data exception - invalid entity type
          </h4>
        </div>
        <div
          class="sc-iwjdpV"
        >
          <pre
            class="sc-cCcXHH haCMIe"
            data-testid="cypherFrameErrorMessage"
          >
            Invalid input '******' for $\`param\`. Expected to be STRING.
          </pre>
        </div>
      </div>
      <div
        class="sc-llYSUQ dMBrER"
      >
        <a
          class="sc-cxpSdN gYcTYD"
        >
           
          Show more
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`ErrorsView does display an error 1`] = `
<div>
  <div
    class="sc-kfPuZi fgEQGn"
  >
    <div
      class="sc-fKVqWL hFvVsk"
    >
      <div
        class="sc-gUQvok ilTLcb"
      >
        <div
          class="sc-bBHxTw cOKOWM"
        >
          <div
            class="sc-iJKOTD sc-ezbkAF hhmYSB fuiMKG"
          >
            ERROR
          </div>
          <h4
            class="sc-hiCibw HoeQn"
          >
            Test.Error
          </h4>
        </div>
        <div
          class="sc-iwjdpV"
        >
          <pre
            class="sc-cCcXHH haCMIe"
            data-testid="cypherFrameErrorMessage"
          >
            Test error description
          </pre>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ErrorsView does display an error for GQL status codes 1`] = `
<div>
  <div
    class="sc-kfPuZi fgEQGn"
  >
    <div
      class="sc-fKVqWL hFvVsk"
    >
      <div
        class="sc-gUQvok ilTLcb"
      >
        <div
          class="sc-bBHxTw cOKOWM"
        >
          <div
            class="sc-iJKOTD sc-ezbkAF hhmYSB fuiMKG"
          >
            ERROR
          </div>
          <h4
            class="sc-hiCibw HoeQn"
          >
            22N14: Data exception - invalid temporal value combination
          </h4>
        </div>
        <div
          class="sc-iwjdpV"
        >
          <pre
            class="sc-cCcXHH haCMIe"
            data-testid="cypherFrameErrorMessage"
          >
            Cannot select both epochSeconds and 'datetime'.
          </pre>
        </div>
      </div>
    </div>
  </div>
</div>
`;
